exonum/src/events/network.rs

```diff
@@ -425,7 +437,28 @@ impl Listener {
         Ok(Listener(to_box(server)))
     }
 
-    fn parse_connect_msg(raw: Option<RawMessage>) -> Result<Connect, failure::Error> {
+    fn handle_incoming_connection(
+        sock: Framed<TcpStream, MessagesCodec>,
+        holder: Weak<()>,
+        address: SocketAddr,
+        network_tx: mpsc::Sender<NetworkEvent>,
+    ) -> impl Future<Item = (), Error = failure::Error> {
+        let (_, stream) = sock.split();
+        stream
+            .into_future()
+            .map_err(|e| e.0)
+            .and_then(|(raw, stream)| (Self::parse_connect_message(raw), Ok(stream)))
+            .and_then(move |(connect, stream)| {
+                trace!("Received handshake message={:?}", connect);
+                Self::process_incoming_messages(stream, network_tx, connect, address)
+            })
+            .map(|_| {
+                // Ensure that holder lives until the stream ends.
+                let _holder = holder;
```

Comment:
By the way, does this actually do anything useful? I don't understand this part.

`holder` seems to be a weak pointer. What's the point of copying it into the closure owned by the future if weak pointers do not extend the lifetime of the owned object? Furthermore, it's `Weak<()>`, a weak pointer to the unit value.

Reply:
Oh, I just noticed the comment explaining the refcount abuse. Yes, using an explicit integer counter is way easier to comprehend than this weak_count thing.

Reply:
I forgot to decrement connections_counter, therefore integer counter is not enough, and I need something like AtomicUsize.

Reply:
It seems that the solution with `Rc` is good because it decrements automatically after the connection is gone. And `holder` ensures that decrement doesn't happen earlier.

Reply:
Actually, the counter does not have to be atomic. Futures are handled on a single thread, and `Rc` uses non-atomic counter as well. The counter can be explicitly decremented when the future completes with something like this:

```diff
-                    .map(|_| {
-                        // Ensure that holder lives until the stream ends.
-                        let _holder = holder;
-                    })
+                    .then(move |result| {
+                        incoming_connections_count -= 1;
+                        result
+                    })
```

Though, you do have a point in that `Rc` does this (semi-)automatically and will probably behave better in case of a panic. But we still need to do this weird dance with making sure that `holder` is copied into the closure and kept around while the future is in progress.

Well, I just found it weird to use `Rc` here. I don't have anything against it.

Reply:
Hm, let's try your approach.

Reply:
Though, I have some doubt for whether using a simple integer will suffice. I'd suppose that it will be moved... _somewhere_ alongside the future, but maybe the compiler will wiggle out and copy the counter which will be incorrect: there will be a copy for `new()` with zero in it, a copy (of zero) for the `ForEach` closure where it is incremented, and another copy of zero for the `Then` closure which will be decremented.

In that case `Rc` does not seem to be a bad choice. After all, `Rc` is used to denote shared ownership and we kinda share the counter between connections. Though, something like `RefCell` may be more semantically correct.

Also, we do not seem to have any tests which verify handling of `max_connections`. This has to be checked manually. If the new approach does not work then I guess we can leave the `Rc` as it, in the hope that it works :) It was there before, after all.

Reply:
I tested manually, integer counter works incorrectly, so I reverted `Rc` again and I think we should stop at this point :sweat: Maybe I return to this issue in the next PR.